mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/scaladsl/MqttSource.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.mqtt.streaming.scaladsl
+
+import akka.actor.ActorSystem
+import akka.{Done, NotUsed}
+import akka.event.{Logging, LoggingAdapter}
+import akka.stream.OverflowStrategy
+import akka.stream.alpakka.mqtt.streaming.impl.Setup
+import akka.stream.alpakka.mqtt.streaming._
+import akka.stream.scaladsl.{BroadcastHub, Flow, Keep, Source, SourceQueueWithComplete, Tcp}
+
+import scala.collection.immutable
+import scala.concurrent.{Future, Promise}
+
+object MqttSource {
+
+  def atMostOnce(settings: MqttSessionSettings,
+                 clientId: String,
+                 subscriptions: immutable.Seq[(String, ControlPacketFlags)]): Source[Publish, Future[Done]] =
+    Setup
+      .source { implicit materializer => implicit attributes =>
+        implicit val system: ActorSystem = materializer.system
+        implicit val logging: LoggingAdapter = Logging.getLogger(system, this)
+
+        val mqttClientSession = ActorMqttClientSession(settings)
+        val mqttFlow: Flow[Command[Nothing], Either[MqttCodec.DecodeError, Event[Nothing]], NotUsed] =
+          Mqtt
+            .clientSessionFlow(mqttClientSession)
+            .join(Tcp(system).outgoingConnection("localhost", 1883))
+
+        val subscribed = Promise[Done]()
+
+        val (commands: SourceQueueWithComplete[Command[Nothing]], subscription: Source[Event[Nothing], NotUsed]) =
+          Source
+            .queue[Command[Nothing]](10, OverflowStrategy.fail)
+            .via(mqttFlow)
+            .log("source received")
+            .map {
+              case Left(decodeError) =>
+                throw new RuntimeException(decodeError.toString)
+// TODO anything we should do if no ConnAck arrives "in time"?
+//            case Right(event @ Event(_: ConnAck, _)) =>
+//              event
+              case Right(event @ Event(_: SubAck, _)) =>
+                subscribed.trySuccess(Done)
+                event
+              case Right(event) =>
+                event
+            }
+            .toMat(BroadcastHub.sink)(Keep.both)
+            .run()
+
+        commands.offer(Command(Connect(clientId, ConnectFlags.CleanSession)))
+        commands.offer(Command(Subscribe(subscriptions)))
+
+        val publishSource: Source[Publish, Future[Done]] = subscription
+          .collect {
+            case Event(publish @ Publish(flags, _, Some(packetId), _), _)
+                if flags.contains(ControlPacketFlags.RETAIN) =>
```

Comment:
Why are you testing for retain? That’s for the server side mostly...

Reply:
Should all `Publish` events trigger a `PubAck` than?
I'll try to let an ack-handle be part of the emitted elements so that acking can be user-controlled for at-least-once semantics.

Reply:
QoS 0 doesn’t require an Ack. QoS 1 does. QoS 2 is quite a dance requiring 2.